// element-plus/lib/el-overlay/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var vnode = require('../utils/vnode');

var script = vue.defineComponent({
    name: 'ElOverlay',
    props: {
        mask: {
            type: Boolean,
            default: true,
        },
        overlayClass: {
            type: String,
        },
        zIndex: {
            type: Number,
        },
    },
    emits: ['click'],
    setup(props, { slots, emit }) {
        const onMaskClick = () => {
            emit('click');
        };
        return () => {
            return props.mask
                ? vue.createVNode('div', {
                    class: ['el-overlay', props.overlayClass],
                    style: {
                        zIndex: props.zIndex,
                    },
                    onClick: onMaskClick,
                }, [vue.renderSlot(slots, 'default')], vnode.PatchFlags.STYLE | vnode.PatchFlags.CLASS | vnode.PatchFlags.PROPS, ['onClick'])
                : vue.renderSlot(slots, 'default');
        };
    },
});

script.__file = "packages/overlay/src/index.vue";

exports.Overlay = script;
