// element-plus/es/el-overlay/index.js
import { defineComponent, createVNode, renderSlot } from 'vue';
import { PatchFlags } from '../utils/vnode';

var script = defineComponent({
    name: 'ElOverlay',
    props: {
        mask: {
            type: Boolean,
            default: true,
        },
        overlayClass: {
            type: String,
        },
        zIndex: {
            type: Number,
        },
    },
    emits: ['click'],
    setup(props, { slots, emit }) {
        const onMaskClick = () => {
            emit('click');
        };
        return () => {
            return props.mask
                ? createVNode('div', {
                    class: ['el-overlay', props.overlayClass],
                    style: {
                        zIndex: props.zIndex,
                    },
                    onClick: onMaskClick,
                }, [renderSlot(slots, 'default')], PatchFlags.STYLE | PatchFlags.CLASS | PatchFlags.PROPS, ['onClick'])
                : renderSlot(slots, 'default');
        };
    },
});

script.__file = "packages/overlay/src/index.vue";

export { script as Overlay };
