// element-plus/es/el-scrollbar/index.js
import { addResizeListener, removeResizeListener } from '../utils/resize-event';
import { toObject } from '../utils/util';
import { defineComponent, getCurrentInstance, ref, inject, computed, onUnmounted, h, provide, onMounted, nextTick, onBeforeUnmount, resolveComponent, openBlock, createBlock, createVNode, resolveDynamicComponent, withCtx, renderSlot, Fragment, createCommentVNode } from 'vue';
import { off, on } from '../utils/dom';

const BAR_MAP = {
    vertical: {
        offset: 'offsetHeight',
        scroll: 'scrollTop',
        scrollSize: 'scrollHeight',
        size: 'height',
        key: 'vertical',
        axis: 'Y',
        client: 'clientY',
        direction: 'top',
    },
    horizontal: {
        offset: 'offsetWidth',
        scroll: 'scrollLeft',
        scrollSize: 'scrollWidth',
        size: 'width',
        key: 'horizontal',
        axis: 'X',
        client: 'clientX',
        direction: 'left',
    },
};
function renderThumbStyle({ move, size, bar }) {
    const style = {};
    const translate = `translate${bar.axis}(${move}%)`;
    style[bar.size] = size;
    style.transform = translate;
    style.msTransform = translate;
    style.webkitTransform = translate;
    return style;
}

var Bar = defineComponent({
    name: 'Bar',
    props: {
        vertical: Boolean,
        size: String,
        move: Number,
    },
    setup(props) {
        const instance = getCurrentInstance();
        const thumb = ref(null);
        const wrap = inject('scroll-bar-wrap', {});
        const bar = computed(() => {
            return BAR_MAP[props.vertical ? 'vertical' : 'horizontal'];
        });
        const barStore = ref({});
        const cursorDown = ref(null);
        const clickThumbHandler = e => {
            if (e.ctrlKey || e.button === 2) {
                return;
            }
            startDrag(e);
            barStore.value[bar.value.axis] = (e.currentTarget[bar.value.offset] - (e[bar.value.client] - e.currentTarget.getBoundingClientRect()[bar.value.direction]));
        };
        const clickTrackHandler = e => {
            const offset = Math.abs(e.target.getBoundingClientRect()[bar.value.direction] - e[bar.value.client]);
            const thumbHalf = (thumb.value[bar.value.offset] / 2);
            const thumbPositionPercentage = ((offset - thumbHalf) * 100 / instance.vnode.el[bar.value.offset]);
            wrap.value[bar.value.scroll] = (thumbPositionPercentage * wrap.value[bar.value.scrollSize] / 100);
        };
        const startDrag = e => {
            e.stopImmediatePropagation();
            cursorDown.value = true;
            on(document, 'mousemove', mouseMoveDocumentHandler);
            on(document, 'mouseup', mouseUpDocumentHandler);
            document.onselectstart = () => false;
        };
        const mouseMoveDocumentHandler = e => {
            if (cursorDown.value === false)
                return;
            const prevPage = barStore.value[bar.value.axis];
            if (!prevPage)
                return;
            const offset = ((instance.vnode.el.getBoundingClientRect()[bar.value.direction] - e[bar.value.client]) * -1);
            const thumbClickPosition = (thumb.value[bar.value.offset] - prevPage);
            const thumbPositionPercentage = ((offset - thumbClickPosition) * 100 / instance.vnode.el[bar.value.offset]);
            wrap.value[bar.value.scroll] = (thumbPositionPercentage * wrap.value[bar.value.scrollSize] / 100);
        };
        function mouseUpDocumentHandler() {
            cursorDown.value = false;
            barStore.value[bar.value.axis] = 0;
            off(document, 'mousemove', mouseMoveDocumentHandler);
            document.onselectstart = null;
        }
        onUnmounted(() => {
            off(document, 'mouseup', mouseUpDocumentHandler);
        });
        return () => h('div', {
            class: ['el-scrollbar__bar', 'is-' + bar.value.key],
            onMousedown: clickTrackHandler,
        }, h('div', {
            ref: thumb,
            class: 'el-scrollbar__thumb',
            onMousedown: clickThumbHandler,
            style: renderThumbStyle({
                size: props.size,
                move: props.move,
                bar: bar.value,
            }),
        }));
    },
});

var script = defineComponent({
    name: 'ElScrollbar',
    components: { Bar },
    props: {
        native: {
            type: Boolean,
            default: false,
        },
        wrapStyle: {
            type: [String, Array],
            default: '',
        },
        wrapClass: {
            type: [String, Array],
            default: '',
        },
        viewClass: {
            type: [String, Array],
            default: '',
        },
        viewStyle: {
            type: [String, Array],
            default: '',
        },
        noresize: Boolean,
        tag: {
            type: String,
            default: 'div',
        },
    },
    setup(props) {
        const sizeWidth = ref('0');
        const sizeHeight = ref('0');
        const moveX = ref(0);
        const moveY = ref(0);
        const wrap = ref(null);
        const resize = ref(null);
        provide('scroll-bar-wrap', wrap);
        const handleScroll = () => {
            if (!props.native && wrap.value) {
                moveY.value = (wrap.value.scrollTop * 100) / wrap.value.clientHeight;
                moveX.value = (wrap.value.scrollLeft * 100) / wrap.value.clientWidth;
            }
        };
        const update = () => {
            if (!wrap.value)
                return;
            const heightPercentage = (wrap.value.clientHeight * 100) / wrap.value.scrollHeight;
            const widthPercentage = (wrap.value.clientWidth * 100) / wrap.value.scrollWidth;
            sizeHeight.value = heightPercentage < 100 ? heightPercentage + '%' : '';
            sizeWidth.value = widthPercentage < 100 ? widthPercentage + '%' : '';
        };
        onMounted(() => {
            if (props.native)
                return;
            nextTick(update);
            !props.noresize && addResizeListener(resize.value, update);
        });
        onBeforeUnmount(() => {
            if (props.native)
                return;
            !props.noresize && removeResizeListener(resize.value, update);
        });
        const style = computed(() => {
            let style = props.wrapStyle;
            if (Array.isArray(props.wrapStyle)) {
                style = toObject(props.wrapStyle);
            }
            return style;
        });
        return {
            moveX,
            moveY,
            sizeWidth,
            sizeHeight,
            style,
            wrap,
            resize,
            update,
            handleScroll,
        };
    },
});

const _hoisted_1 = { class: "el-scrollbar" };

function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_bar = resolveComponent("bar");

  return (openBlock(), createBlock("div", _hoisted_1, [
    createVNode("div", {
      ref: "wrap",
      class: [
        _ctx.wrapClass,
        'el-scrollbar__wrap',
        _ctx.native ? '' : 'el-scrollbar__wrap--hidden-default',
      ],
      style: _ctx.style,
      onScroll: _cache[1] || (_cache[1] = (...args) => (_ctx.handleScroll(...args)))
    }, [
      (openBlock(), createBlock(resolveDynamicComponent(_ctx.tag), {
        ref: "resize",
        class: ['el-scrollbar__view', _ctx.viewClass],
        style: _ctx.viewStyle
      }, {
        default: withCtx(() => [
          renderSlot(_ctx.$slots, "default")
        ]),
        _: 3
      }, 8 /* PROPS */, ["class", "style"]))
    ], 38 /* CLASS, STYLE, HYDRATE_EVENTS */),
    (!_ctx.native)
      ? (openBlock(), createBlock(Fragment, { key: 0 }, [
          createVNode(_component_bar, {
            move: _ctx.moveX,
            size: _ctx.sizeWidth
          }, null, 8 /* PROPS */, ["move", "size"]),
          createVNode(_component_bar, {
            vertical: "",
            move: _ctx.moveY,
            size: _ctx.sizeHeight
          }, null, 8 /* PROPS */, ["move", "size"])
        ], 64 /* STABLE_FRAGMENT */))
      : createCommentVNode("v-if", true)
  ]))
}

script.render = render;
script.__file = "packages/scrollbar/src/index.vue";

script.install = (app) => {
    app.component(script.name, script);
};
const _Scrollbar = script;

export default _Scrollbar;
